Reorganize ray_deps_setup.bzl to make all the GitHub rules uniform and download ZIP files for everything

mehrdadn opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

- ZIP file downloads are much faster than git clones, due to the lack of history, resulting in much faster build times
- The build rules have redundancy and are becoming disorganized and hard to maintain

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

mehrdadn commented:
@jovany-wang Yes, it's necessary. `http_archive` has been a source of redundancy and disorganization, and previously there has been a need for `git_repository` as well.

Compare what it was like before the change to what it's like after. Before the change, we had:

    RULES_JVM_EXTERNAL_TAG = "2.10"
    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
    http_archive(
        name = "rules_jvm_external",
        sha256 = RULES_JVM_EXTERNAL_SHA,
        strip_prefix = "rules_jvm_external-%s" % RULES_JVM_EXTERNAL_TAG,
        url = "https://github.com/bazelbuild/rules_jvm_external/archive/%s.zip" % RULES_JVM_EXTERNAL_TAG,
    )

Now we have:

    github_repository(
        name = "rules_jvm_external",
        tag = "2.10",
        remote = "https://github.com/bazelbuild/rules_jvm_external",
        sha256 = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26",
    )

jovany-wang commented:
@mehrdadn  I see.  It simplified the process of catting remote address and version-tag to url.

mehrdadn commented:
@jovany-wang Yeah-- and also simplified the process of switching between git and zip and local repos, and also removed the redundancy in `build_file` and `strip_prefix`, and also made it easier to test locally with `path =` (fewer merge conflicts)... it simplified a fair number of things I think. :-)

jovany-wang commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+def github_repository(**kwargs):
+    """
+    Conveniently chooses between ZIP, git, etc. GitHub repositories.
+    We prefer ZIP downloads, since they're smaller and faster
```
```suggestion
    Recommend to download zip packages, because zip packages have no history of the repositories.
```

jovany-wang replied:
BTW, the comments of this method seems too verbose, could you refine them?

mehrdadn replied:
I agree it's more verbose for my liking too, but I've in fact already rewritten it several times to get it into this shape. :) I removed the parameters that are already described in Bazel and only documented any additional knowledge needed. I'm not sure what else to condense without removing pertinent information to be honest. I'm happy to consider an alternative if you have an suggestions on how to shorten it, though.

jovany-wang commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
-def ray_deps_setup():
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
```
How about naming it `ray_dependency_repository()`, because we use this method not only declaring github dependencies, but also declaring other type dependencies.

mehrdadn replied:
I'm not sure what you mean -- what other type dependencies is this used for? Note that this is intended strictly for GitHub, and it's not specific to Ray in any way -- you can't in general switch between ZIP archives and git clones; it's only possible on GitHub because they have a well-defined URL format.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18584/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18589/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18594/
Test PASSed.

edoakes commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+            kwargs.setdefault("shallow_since", shallow_since)
+            kwargs.setdefault("remote", remote)
+            if "build_file" in kwargs:
+                new_git_repository(**kwargs)
```
please just call this and `git_repository` below with expanded keyword arguments for clarity, otherwise I have to trace back all of the code that modifies `kwargs` to understand what this line is passing in

edoakes replied:
and `http_archive` below

mehrdadn replied:
Hmm okay, if you confirm https://github.com/ray-project/ray/pull/6193/files#r348190605 then I'll do this.

mehrdadn replied:
Update: I think this isn't workable anymore in light of https://github.com/ray-project/ray/pull/6193#discussion_r348218943.

edoakes commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+    are automatically stripped when not applicable.
+    """
+
+    remote = kwargs.pop("remote")  # required
```
if this is required it shouldn't be a kwarg

mehrdadn replied:
We don't want it to be positional. That's not normal Bazel practice for dependency rules. It would also interfere with the canonical ordering of arguments.

edoakes replied:
In that case can we just check that it's set and raise a useful error?

mehrdadn replied:
Ok

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+
+    If build_file is a string, it refers to a custom build file.
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
```
@edoakes Yeah, it's a lot of redundancy for no gain. It doesn't affect reproducibility in any way.

mehrdadn replied:
Yeah, it's a lot of redundancy for no gain. It doesn't affect reproducibility in any way.

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+    """
+    Conveniently chooses between ZIP, git, etc. GitHub repositories.
+    We prefer ZIP downloads, since they're smaller and faster
+    due to the lack of history.
```
We're not calling git manually here, Bazel is calling it. And it performs a lot worse than archive downloads.

edoakes replied:
Looks like bazel calls with depth=1 if `tag` is specified:
https://docs.bazel.build/versions/0.25.0/repo/git.html

mehrdadn replied:
Tags don't have any integrity checking though; clones do.
Also, I saw maybe 2 tags out of the numerous dependencies we have; everything else was by commit hash. I don't think you can realistically assume every commit we want from every dependency will necessarily have a tag for us to use.

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+
+    If path   is present , local repository is assumed.
+    If sha256 == <string>, obtains ZIP download with the given hash (fast)
+    If sha256 == True    , obtains ZIP download without knowing the hash (fast)
```
Yes, as follows:

    Set sha256 = True if the ZIP archive hash is unknown, then update it ASAP
    to allow caching & avoid repeated downloads on every build.

edoakes replied:
Is it used by any of our build targets though? No point in adding it if not and this seems like an easy way to accidentally end up with slow builds.

mehrdadn replied:
No what I mean is that it should never be checked in (similar to how `local` will be checked in), but it is used locally to download the ZIP without knowing the hash a priori. And as a bonus Bazel tells you what the hash is when you do that, so it's quite easy to just copy-paste from the terminal that's already in front of you. Otherwise you have force everyone (including yourself) to manually download the ZIP in the browser and run `sha256sum` before you can add a dependency at all. And then you have to have Bazel download it one _more_ time now that it has the hash. Which is both quite painful in itself, and also a time-waster given that some ZIPs are fairly large and now have to be downloaded twice.

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+
+    remote = kwargs.pop("remote")  # required
+
+    GIT_SUFFIX = ".git"  # We need to strip this
```
That's wrong though. `"foobartig.git".rstrip(".git") == "foobar"`

edoakes replied:
wow, that is an insane design choice

mehrdadn replied:
Yeah. I think the reason they allow it to be a string is for efficiency, but I think they should've forced it to be a keyword argument if the caller wanted to provide more than one character.

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
-def ray_deps_setup():
-    RULES_JVM_EXTERNAL_TAG = "2.10"
-    RULES_JVM_EXTERNAL_SHA = "1bbf2e48d07686707dd85357e9a94da775e1dbd7c464272b3664283c9c716d26"
+def github_repository(**kwargs):
```
I can if you really want to, but are do you really prefer having to update this function every time someone wants to use an argument that wasn't previously in the supported list? I tried to make this transparent so it would just pass unknown arguments through.

edoakes replied:
Yes that would be preferred

mehrdadn replied:
Ok I'll change it

mehrdadn replied:
@edoakes Wait I just realized I can't do that, because if I put them here then people will try to call them positionally rather than as keywords. And Starlark (which is this Python-like language) forbids keyword-only arguments, so I can't do that either.

That said though, I'll try to shuffle the logic to make it easier to see what the defaults are.

mehrdadn replied:
Update: see new commits.

mehrdadn commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,128 +1,198 @@
+    If build_file is absent, the project is assumed to contain a build file.
+    If build_file == True, then the build file name is auto-deduced.
+
+    If path   is present , local repository is assumed.
```
Yes, I use it for development (as can anyone who may want to try patching a dependency). It's for local copies of repositories, so it of course doesn't make sense to be used for checked-in code.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18614/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18613/
Test FAILed.

edoakes commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,131 +1,194 @@
+    sha256 should be omitted (or None) when the archive hash is unknown, then
+    updated ASAP to allow caching & avoid repeated downloads on every build.
+
+    NOTE: False and None mean different things.
```
don't need this, it's a python convention :)

mehrdadn replied:
Alrighty :P

edoakes commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,131 +1,194 @@
+
+    treeish = commit or tag or branch
+    if not treeish: fail("Missing commit, tag, or branch argument")
+    if remote == None: fail("Missing remote argument")
```
`if remote is None`

mehrdadn replied:
@edoakes I wish. Starlark requires `==`. :(

edoakes commented on bazel/ray_deps_setup.bzl:
```diff
@@ -1,131 +1,194 @@
+
+    project = (remote[:-len(GIT_SUFFIX)]  # We need to strip this suffix
+               if remote.endswith(GIT_SUFFIX)
+               else remote).split("//", 1)[1].split("/")[2]
```
This is pretty hard to read, would be nice if you could you separate out the logic a bit into a few lines and add some comments about how exactly we're mangling this string

mehrdadn replied:
Done

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18719/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18728/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18743/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18745/
Test FAILed.
